### PR TITLE
cln-plugin: Save "configuration" from "init" method

### PR DESCRIPTION
**Pull request by justinmoon**
`cln-plugin` throws away the "configuration" field from the "init" message.

### CODE REVIEW

File: plugins/src/messages.rs [hunk 1]
```diff
@@ -63,6 +64,8 @@ pub struct GetManifestCall {}
 #[derive(Deserialize, Debug)]
 pub(crate) struct InitCall {
     pub(crate) options: HashMap<String, Value>,
+    pub(crate) configuration: Value,
+    // pub(crate) configuration: Configuration,
```
Comment (justinmoon) [inline plugins/src/messages.rs:68]: If I try to put `Configuration` in here, then [this match](https://github.com/ElementsProject/lightning/blob/master/plugins/src/lib.rs#L188) isn't triggered. In principle it seems like this should work if everything is `Deserialize` ...

**Comment by justinmoon** [comment]
Added a unittest as separate commit to make sure the "init" message (used [this example message](https://github.com/ElementsProject/lightning/blob/master/doc/PLUGINS.md#the-init-method)) can be parsed. Not sure if this is helpful or not.

File: plugins/src/messages.rs [hunk 2]
```diff
@@ -63,6 +64,7 @@ pub struct GetManifestCall {}
 #[derive(Deserialize, Debug)]
 pub(crate) struct InitCall {
     pub(crate) options: HashMap<String, Value>,
+    pub(crate) configuration: Configuration,
```
Comment (justinmoon) [inline plugins/src/messages.rs:67]: Would you prefer I just use `Value` here?

File: plugins/src/lib.rs [hunk 3]
```diff
@@ -653,6 +662,12 @@ where
     }
 }
 
+#[derive(Clone, Debug, Deserialize)]
+pub struct Configuration {
+    #[serde(rename = "lightning-dir")]
+    pub lightning_dir: String,
+}
```
Comment (justinmoon) [inline plugins/src/lib.rs:669]: If we keep this struct, what else should go in it?

File: plugins/src/messages.rs (hunk 2 above)
Comment (vincenzopalazzo) [reply plugins/src/messages.rs:67]: Maybe an HasMap<String, Value> or keep the Configuration here! But if I'm not wrong @cdecker wants this library strongly typed, perhaps the struct is better?

File: plugins/src/lib.rs (hunk 3 above)
Comment (vincenzopalazzo) [reply plugins/src/lib.rs:669]: Here the docs https://lightning.readthedocs.io/PLUGINS.html#the-init-method and if you want to do a copy and paste https://github.com/laanwj/rust-clightning-rpc/blob/92912c8a5318624b2c9684676a0f697b4a38432e/plugin/src/commands/types.rs#L25-L51

File: plugins/src/lib.rs [hunk 4]
```diff
@@ -363,7 +369,7 @@ where
     /// The state gets cloned for each request
     state: S,
     options: Vec<ConfigOption>,
-
+    configuration: Configuration,
```
Comment (vincenzopalazzo) [inline plugins/src/lib.rs:372]:
```suggestion
    /// The cln configuration received after the init msg
    configuration: Configuration,
```

**Review by vincenzopalazzo — CHANGES_REQUESTED** [review]
LGTM, just take a decision if we want use an HasMap or a Struct for the configuration

**Comment by cdecker** [comment]
> overall LGTM, just take a decision if we want use an HasMap or a Struct for the configuration

As you mentioned above making these things we have control over as strongly typed as possible would be a good idea. So I'd vote in favor of the `Configuration` struct 👍

File: plugins/src/lib.rs (hunk 4 above)
Comment (justinmoon) [reply plugins/src/lib.rs:376]: I commented this key and the "options" key

File: plugins/src/messages.rs (hunk 2 above)
Comment (justinmoon) [reply plugins/src/messages.rs:66]: Went with "configuration". I prefer that too.

**Comment by justinmoon** [comment]
Added the rest of the fields to `Configuration` mostly copying from [rust-clightning-rpc](https://github.com/laanwj/rust-clightning-rpc/blob/92912c8a5318624b2c9684676a0f697b4a38432e/plugin/src/commands/types.rs#L25-L51) as suggested by @vincenzopalazzo. I prefer the strongly typed struct too.

**Comment by vincenzopalazzo** [comment]
It is missing the Changelog in the body commit, maybe `Changelog-Added: cln_plugin: take into consideration the cln configuration from init msg` can be a good description?

**Comment by justinmoon** [comment]
Added changelog entry @vincenzopalazzo

**Comment by rustyrussell** [comment]
BTW I've invited you to project as a reader. This should mean CI runs for you...

File: plugins/src/messages.rs [hunk 5]
```diff
@@ -63,6 +63,30 @@ pub struct GetManifestCall {}
...
+    pub proxy: ProxyInfo,
+    #[serde(rename = "torv3-enabled")]
+    pub torv3_enabled: bool,
+    pub always_use_proxy: bool,
+}
+
+#[derive(Clone, Debug, Deserialize)]
+pub struct ProxyInfo {
+    #[serde(alias = "type")]
+    pub tup: String,
```
Comment (vincenzopalazzo) [inline plugins/src/messages.rs:87]:
```suggestion
    pub typ: String,
```

This is my typo, sorry!

**Review by vincenzopalazzo — CHANGES_REQUESTED** [review]
Only a small comment, and LGTM.

With the next commit, we will see with the CI will be happy!

File: plugins/src/messages.rs (hunk 5 above)
Comment (justinmoon) [reply plugins/src/messages.rs]: Done

**Comment by justinmoon** [comment]
The tests seem to be failing because [this match is failing](https://github.com/ElementsProject/lightning/blob/9a880a0932e7d8575620a1f7b48038c52b4b126f/plugins/src/lib.rs#L187) with my `Configuration` struct. But AFAIKT [my unittest does exactly the same match](https://github.com/ElementsProject/lightning/pull/5279/files#diff-28222d09a6ae0658f211b675fe20ca53e5fde769dcfeef8e3b181123e26951b2R205). 

In my unittest I'm using [exactly this JSON](https://github.com/ElementsProject/lightning/blob/master/doc/PLUGINS.md#the-init-method) and it passes. Why is it failing in the python test in CI?

It does seem that the example JSON in the link above is different from what the CI test is sending. For example, the `always_use_proxy` value, for example, is [missing in CI](https://github.com/ElementsProject/lightning/runs/6965889120?check_suite_focus=true#step:5:2902).

**Comment by cdecker** [comment]
ACK https://github.com/ElementsProject/lightning/pull/5279/commits/9e912f85666a7ec8929544342e6952e58b1352a0

**Pull request merged**